android-app/app/src/main/java/de/dhbw/ka/se/fibo/ui/dashboard/DashboardFragment.java

```diff
@@ -147,20 +207,67 @@ private void initializePieChart() {
         pieChart.animateY(DashboardFragment.PIE_CHART_ANIMATION_DURATION, Easing.EaseInOutQuad);
     }
 
+    /**
+     * Loads all cashflows from the application state, extracts all expenses and applies all active filters
+     *
+     * @return the expenses per category to be displayed to the user
+     */
+    @NonNull
+    private Map<Category, BigDecimal> getFilteredExpensesPerCategory() {
+        Context context = requireContext();
+
+        SortedSet<Cashflow> cashflows = ApplicationState.getInstance(context).getCashflows();
+
+        Stream<Cashflow> cashflowStream = cashflows.stream().filter(x -> CashflowType.EXPENSE == x.getType());
+
+        // filter by start date
+        if (null != startDate) {
+            cashflowStream = cashflowStream.filter(x -> startDate.minusDays(1).isBefore(ChronoLocalDate.from(x.getTimestamp())));
+        }
+
+        // filter by end date
+        if (null != endDate) {
+            cashflowStream = cashflowStream.filter(x -> endDate.plusDays(1).isAfter(ChronoLocalDate.from(x.getTimestamp())));
+        }
+
+        Map<Category, BigDecimal> expensesPerCategory = new HashMap<>();
+
+        for (Iterator<Cashflow> it = cashflowStream.iterator(); it.hasNext(); ) {
+            Cashflow cashflow = it.next();
+
+            Category category = cashflow.getCategory();
+
+            // filter by category
+            if (hiddenCategories.contains(category)) {
+                continue;
+            }
+
+            BigDecimal newValue = expensesPerCategory
+                    .computeIfAbsent(category, x -> BigDecimal.ZERO)
+                    .add(cashflow.getOverallValue());
+
+            expensesPerCategory.put(category, newValue);
+        }
+
+        return expensesPerCategory;
+    }
+
```

Comment:
what about `cashflowStream.forEach(cashflow ->{})` for a better overview?

Reply:
yup, we could do that!

Reply:
done :)